### PR TITLE
Allow blocking in message handling code

### PR DESCRIPTION
**Pull request by kasperl**

### CODE REVIEW

File: tests/max_heap_size_test.toit [hunk 1]
```diff
@@ -17,7 +17,12 @@ expect_allocation_failed [code]:
 main:
   // We can limit ourselves to as little as a 4k heap (on 32 bit) which
   // means no old-space.
-  for i := 12; i < 17; i++:
+
+  // TODO(kasper): Bumped lower limit 12 (4k) to 13 (8k) to unblock work
```
Comment (kasperl) [inline tests/max_heap_size_test.toit:21]: Will check if this is still necessary given my latest changes.

File: lib/core/message.toit [hunk 2]
```diff
@@ -88,23 +93,82 @@ process_messages_:
...
+              print_ "WARNING: unhandled system message $type"
+
+          reused := true
+          processor := message_processor_
+          if not processor.run lambda:
+            processor = MessageProcessor_ lambda
+            message_processor_ = processor
+            reused = false
+          processor.detach_if_necessary_
+          if not entered:
```
Comment (kasperl) [inline lib/core/message.toit:117]: Will remove this code before merging. It is here to make any issue with this stand out more during testing.

File: lib/core/message.toit [hunk 3]
```diff
@@ -88,23 +93,82 @@ process_messages_:
...
+          entered := false
+          lambda := ::
+            entered = true
+            type ::= message[0]
+            handler ::= system_message_handlers_.get type
+            if handler:
+              gid ::= message[1]
+              pid ::= message[2]
+              arguments ::= message[3]
+              message = null  // Allow garbage collector to free.
```
Comment (floitsch) [inline lib/core/message.toit]: Hopefully shouldn't be necessary in the future when the variable slots are reused.
Also not sure it buys a lot, but can't really hurt.

File: lib/core/message.toit [hunk 4]
```diff
@@ -88,23 +93,74 @@ process_messages_:
...
+              handler.on_message type gid pid arguments
+            else:
+              print_ "WARNING: unhandled system message $type"
+          processor := message_processor_
+          if not processor.run lambda:
+            processor = MessageProcessor_ lambda
```
Comment (floitsch) [inline lib/core/message.toit]:
```suggestion
            // The processor has already a lambda or doesn't have a task anymore.
            // Simply create a new one and replace the old one.
            processor = MessageProcessor_ lambda
```

File: lib/core/finalization.toit [hunk 5]
```diff
@@ -47,20 +47,27 @@ monitor FinalizationStack_:
...
 
-  wait_for_next -> bool:
+  wait_for_next -> Lambda?:
     deadline := Time.monotonic_us + IDLE_TIME_MS * 1_000
-    return try_await --deadline=deadline: not lambdas_.is_empty
+    try_await --deadline=deadline: not lambdas_.is_empty
+    // If we got a lambda, we must return it and deal with even if we timed out.
+    return lambdas_.is_empty ? null : lambdas_.remove_last
 
   ensure_finalization_task_ -> none:
```
Comment (floitsch) [inline lib/core/finalization.toit:56]: Name feels a bit off. Sounds like it's a field/getter for a task. However, it is running/creating something instead.

**Review by floitsch — APPROVED** [review]
LGTM.

**Review by floitsch — APPROVED** [review]
Still LGTM.

File: lib/core/message.toit (hunk 3 above)
Comment (kasperl) [reply lib/core/message.toit]: Dropped this in another change.

File: lib/core/finalization.toit (hunk 5 above)
Comment (kasperl) [reply lib/core/finalization.toit:56]: It's a valid point.

**Pull request merged**